target/linux/bmips/image/Makefile

```diff
@@ -0,0 +1,262 @@
+	# vmlinux is located on a different JFFS2 partition, but CFE RAM
+	# ignores it, so let's create another dummy file that will match
+	# the JFFS2 ino of cferam entry on the first JFFS2 partition.
+	# CFE RAM won't be able to find vmlinux if cferam has the same
+	# ino as vmlinux.
+	touch $@-kernel/2-openwrt
+	# Add vmlinux as the last file in the JFFS2 partition
+	$(TOPDIR)/scripts/cfe-bin-header.py \
+		--input-file $@ \
+		--output-file $@-kernel/vmlinux.lz \
+		--load-addr $(if $(DEVICE_LOADADDR),$(DEVICE_LOADADDR),$(LOADER_ENTRY)) \
+		--entry-addr $(if $(DEVICE_LOADADDR),$(DEVICE_LOADADDR),$(LOADER_ENTRY))
+
+	# The JFFS2 partition creation should result in the following
+	# layout:
+	# 1) 1-openwrt (version 0, ino 2)
+	# 2) 2-openwrt (version 1, ino 3)
+	# 3) vmlinux.lz (version 2, ino 4)
+	$(call Build/cfe-jffs2,$@-kernel)
+endef
+
+define Build/cfe-part-tag
+	mv $@ $@.part
+
+	$(TOPDIR)/scripts/cfe-partition-tag.py \
+		--input-file $@.part \
+		--output-file $@ \
+		--flags $(CFE_PART_FLAGS) \
+		--id $(CFE_PART_ID) \
+		--name $(VERSION_CODE) \
+		--version $(DEVICE_NAME)
+
+	$(call Build/pad-to,$(BLOCKSIZE))
+
+	dd if=$@.part >> $@
+endef
+
+define Build/cfe-sercomm-part
+	$(TOPDIR)/scripts/sercomm-partition-tag.py \
+		--input-file $@ \
+		--output-file $@.kernel_rootfs \
+		--part-name kernel_rootfs \
+		--part-version OpenWrt \
+		--rootfs-version $(SERCOMM_VERSION)
+
+	rm -rf $@-rootfs_lib
+	mkdir -p $@-rootfs_lib
+	echo $(SERCOMM_VERSION) > $@-rootfs_lib/lib_ver
+	$(call Build/cfe-jffs2,$@-rootfs_lib)
+	$(call Build/pad-to,$(BLOCKSIZE))
+	$(TOPDIR)/scripts/sercomm-partition-tag.py \
+		--input-file $@ \
+		--output-file $@.rootfs_lib \
+		--part-name rootfs_lib \
+		--part-version $(SERCOMM_VERSION)
+
+	mv $@.kernel_rootfs $@
+	dd if=$@.rootfs_lib >> $@
+endef
+
+define Build/cfe-sercomm-load
+	$(TOPDIR)/scripts/sercomm-payload.py \
+		--input-file $@ \
+		--output-file $@.new \
+		--pid "$(SERCOMM_PID)"
+
+	mv $@.new $@
+endef
+
+define Build/cfe-sercomm-crypto
+	$(TOPDIR)/scripts/sercomm-crypto.py \
+		--input-file $@ \
+		--key-file $@.key \
+		--output-file $@.ser \
+		--version OpenWrt
+	$(STAGING_DIR_HOST)/bin/openssl enc -md md5 -aes-256-cbc \
+		-in $@ -out $@.enc \
+		-K `cat $@.key` \
+		-iv 00000000000000000000000000000000
+	dd if=$@.enc >> $@.ser
+	mv $@.ser $@
+	rm -f $@.enc $@.key
+endef
+
+define Build/cfe-wfi-tag
+	$(TOPDIR)/scripts/cfe-wfi-tag.py \
+		--input-file $@ \
+		--output-file $@.new \
+		--version $(if $(1),$(1),$(CFE_WFI_VERSION)) \
+		--chip-id $(CFE_WFI_CHIP_ID) \
+		--flash-type $(CFE_WFI_FLASH_TYPE) \
+		$(if $(CFE_WFI_FLAGS),--flags $(CFE_WFI_FLAGS))
+	mv $@.new $@
+endef
+
+### Device scripts ###
+define Device/Default
+  PROFILES = Default $$(DEVICE_NAME)
+  KERNEL_DEPENDS = $$(wildcard ../dts/$$(DEVICE_DTS).dts)
+  DEVICE_DTS_DIR := ../dts
+  CHIP_ID :=
+  SOC = bcm$$(CHIP_ID)
+  DEVICE_DTS = $$(SOC)-$(subst _,-,$(1))
+  DEVICE_LOADADDR :=
+endef
+DEVICE_VARS += CHIP_ID DEVICE_LOADADDR
```

Comment:
Consider moving DEVICE_VARS to the top, like in the *.mk files or here:
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/image/Makefile#L6